### DISCUSSION TITLE
Diagonal line in table cell

### DISCUSSION THREAD

**Discussion by bugsbugsbux** [typst/typst#4800] [Q&A] [answered]
How do I draw a diagonal line in a table cell from one corner to another?

Thanks!

**Comment by PgBiel** [comment]
You can use this library: https://github.com/PgBiel/typst-diagbox

I've been meaning to update and clean it up for publishing for a while now, but haven't had the opportunity yet, sadly. So you'll have to copy the code to your project (it's a single file).

**Reply by bugsbugsbux** [reply]
interesting, but does not seem to work if there shall be no text in the cell and only the line...

**Reply by PgBiel** [reply]
Sorry, what do you mean? You can use [] to indicate empty content

What code did you try and what was the result?

**Reply by bugsbugsbux** [reply]
I tried:
```
#import "diagbox.typ": *

#table(columns: 2,
    bdiagbox[][], [foo],
    [bar],
)
```

The result is the table with a short line sticking into the first cell from its upper left corner to the cell's middle in height and a bit left of its middle in width.

**Reply by PgBiel** [reply, marked as answer]
Weird, seems like a bug. For now, you can use some invisible text as a workaround, for example:

```typ
#table(columns: 2,
    bdiagbox(hide[aa], hide[bb]), [foo],
    [bar],
)
```

`bdiagbox` also has several settings which may be useful.

**Comment by hott-onin-1** [comment]
I can suggest a more streamlined approach with more recent Typst:

```typst
#let diagbox(text_left, text_right, padding: 5pt, stroke: 1pt, inverted: false) = context {    
  let padded_right = pad(text_right, padding)
  let padded_left = pad(text_left, padding)
  
  let measure_right = measure(padded_right)
  let measure_left = measure(padded_left)
  
  let width_diff = calc.abs(measure_right.width - measure_left.width)
  
  let inner_height = measure_right.height + measure_left.height + width_diff / 10
  let inner_width = measure_right.width + measure_left.width

  box(width: inner_width, height: inner_height) // Empty box to ensure minimal size
  
  if inverted {
    place(bottom + right, padded_right)
    place(top + left, line(start: (0%, 100%), end: (100%, 0%), stroke: stroke))
    place(top + left, padded_left)
  }
  else {
    place(top + right, padded_right)
    place(top + left, line(end: (100%, 100%), stroke: stroke))
    place(bottom + left, padded_left)
  }
}

#table(
  columns: 2,
  align: center + horizon,
  table.cell(diagbox()[A][B], inset: 0pt, breakable: false), [],
  [], [],
)
```

It worked better for my use cases than [typst-diagbox](https://github.com/PgBiel/typst-diagbox), especially with various text sizes and various cell sizes in the rest of the table. It doesn't rely too much on manual sizing / calculation but require `inset: 0pt, breakable: false` in the cell using the diagbox.

**Reply by PgBiel** [reply]
Yeah, that is the natural evolution - I had diagbox in mind when proposing per-cell inset configuration, but never got to properly recommend it. Perhaps you can submit a PR to the diagbox repo?

**Reply by PgBiel** [reply]
I also think you might still need to use fixed size columns for full predictability of the diagbox size due to #4000

**Reply by hott-onin-1** [reply]
Yeah I thought about submitting a PR, but I'm still not happy with the  function requiring to be wrapped in a `table.cell` with `inset: 0pt, breakable: false` and I couldn't manage to abstract it. (also it would be breaking change but that's another problem)

**Reply by hott-onin-1** [reply]
Never-mind I made it work, `context` was messing with me:

```typst
/**
 * A diagonal box for table headings.
 *
 * 'text_left' is the text that appears on the left (top or bottom).
 * 'text_right' is the text that appears on the right (top or bottom).
 * 'start' controls where the diagonal will start: `top` will create a line from top left to bottom right, `bottom` will create a line from bottom left to top right. Defaults to `top`.
 * Additionaly, the function accepts all parameters from `table.cell`
 */
#let diagbox(text_left, text_right, start: "top", ..kwargs) = {
  let content = context {

    let stroke = kwargs.at("stroke", default: 1pt + black)
    let inset = kwargs.at("inset", default: 0% + 5pt)
    
    let padded_right = pad(text_right, inset)
    let padded_left = pad(text_left, inset)

    let measure_right = measure(padded_right)
    let measure_left = measure(padded_left)

    // Used to account for big differences between text widths
    let width_diff = calc.abs(measure_right.width - measure_left.width)

    let inner_height = measure_right.height + measure_left.height + width_diff / 10
    let inner_width = measure_right.width + measure_left.width

    box(width: inner_width, height: inner_height) // Empty box to ensure minimal size

    if start == "bottom" {
      place(bottom + right, padded_right)
      place(top + left, line(start: (0%, 100%), end: (100%, 0%), stroke: stroke))
      place(top + left, padded_left)
    }
    else if start == "top" {
      place(top + right, padded_right)
      place(top + left, line(end: (100%, 100%), stroke: stroke))
      place(bottom + left, padded_left)
    }
    else {
      panic("Unhandled value '" + start + "' for start parameter, expected 'top' or 'bottom'")
    }
  }
  table.cell(content, ..kwargs, inset: 0pt, breakable: false)
}

#table(
  columns: 2,
  align: center + horizon,
 diagbox()[A][B], [],
  [], [],
)
```

**Reply by PgBiel** [reply]
We can't return a table.cell directly because you may also need to use it in a grid, then you'd return grid.cell. A better idea would be to have separate functions like table-(b/t)diagbox and grid-(b/t)diagbox, or even just have a parameter for that.
